File: src/main/java/waterpower/common/block/reservoir/TileEntityReservoir.java
package waterpower.common.block.reservoir;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidRegistry;
import net.minecraftforge.fluids.FluidStack;

import org.apache.commons.lang3.mutable.MutableObject;

import net.minecraftforge.fml.common.Optional.Interface;
import net.minecraftforge.fml.common.Optional.InterfaceList;
import waterpower.Reference;
import waterpower.WaterPower;
import waterpower.api.IUpgrade;
import waterpower.api.IWaterReceiver;
import waterpower.client.gui.DefaultGuiIds;
import waterpower.common.block.inventory.InventorySlotConsumableLiquid;
import waterpower.common.block.inventory.InventorySlotOutput;
import waterpower.common.block.inventory.InventorySlotUpgrade;
import waterpower.common.block.tileentity.TileEntityMultiBlock;
import waterpower.util.Mods;
import waterpower.util.Pair;
import waterpower.util.Utils;

@InterfaceList({ @Interface(iface = "ic2.api.tile.IWrenchable", modid = Mods.IDs.IndustrialCraft2API, striprefs = true) })
public class TileEntityReservoir extends TileEntityMultiBlock {

    public EnumFacing side;
    public Reservoir size = new Reservoir();

    private int lastLength = -1, lastWidth = -1, lastHeight = -1;
    private double rainLevel, underLevel, overLevel;
    private int lastAddedWater = 0, preLastAddedWater = -999;
    private Pair ocean = new Pair(0, 0);
    boolean sendInitData;

    private final InventorySlotConsumableLiquid fluidSlot = new InventorySlotConsumableLiquid(this, "fluidSlot", 1);
    private final InventorySlotOutput outputSlot = new InventorySlotOutput(this, "output", 1);
    private final InventorySlotUpgrade upgradeSlot = new InventorySlotUpgrade(this, "upgrade", 4);
    private int defaultStorage, extraStorage, storage;

    public TileEntityReservoir() {
    	super(0);
    }
    
    public TileEntityReservoir(ReservoirType type) {
    	super(type.capacity);
    	temporaryType = type;
    }

    @Override
    public void readPacketData(NBTTagCompound tag) {
        super.readPacketData(tag);

        lastAddedWater = tag.getInteger("lastAddedWater");
        size.width = tag.getInteger("reservoirWidth");
        size.length = tag.getInteger("reservoirLength");
        size.height = tag.getInteger("reservoirHeight");
        size.valid = tag.getBoolean("reservoirValid");
    }

    @Override
    public void writePacketData(NBTTagCompound tag) {
        super.writePacketData(tag);

        tag.setInteger("lastAddedWater", lastAddedWater);
        
        tag.setInteger("reservoirWidth", size.width);
        tag.setInteger("reservoirLength", size.length);
        tag.setInteger("reservoirHeight", size.height);
        tag.setBoolean("reservoirValid", size.valid);
    }

    public InventorySlotConsumableLiquid getFluidSlot() {
        if (isMaster())
            return fluidSlot;
        else
            return ((TileEntityReservoir) masterBlock).fluidSlot;
    }

    public InventorySlotOutput getOutputSlot() {
        if (isMaster())
            return outputSlot;
        else
            return ((TileEntityReservoir) masterBlock).outputSlot;
    }

    public InventorySlotUpgrade getUpgradeSlot() {
        if (WaterPower.isClientSide() || isMaster())
            return upgradeSlot;
        else
            return ((TileEntityReservoir) masterBlock).upgradeSlot;
    }
    
    ReservoirType temporaryType = null;
    
    public ReservoirType getType() {
    	if (temporaryType != null) return temporaryType;
    	return temporaryType = ReservoirType.values()[getBlockMetadata()];
    }

    @Override
    public String getName() {
        return getType().getShowedName();
    }

    public void useLiquid(int use) {
        if (WaterPower.isClientSide() || isMaster())
            fluidTank.drain(use, true);
        else if (masterBlock != null)
            ((TileEntityReservoir) masterBlock).fluidTank.drain(use, true);
    }

    @Override
    protected boolean canBeMaster() {
        if (Reservoir.isRes(worldObj, pos.down(), getType()) || Reservoir.isRes(worldObj, pos.west(), getType())
                || Reservoir.isRes(worldObj, pos.north(), getType())) {
            return false;
        }
        return true;
    }

    private ArrayList<TileEntityMultiBlock> setSide(ArrayList<TileEntityMultiBlock> a, EnumFacing d) {
        for (TileEntityMultiBlock block : a) {
            TileEntityReservoir te = (TileEntityReservoir) block;
            te.side = d;
        }
        return a;
    }

    @Override
    protected ArrayList<TileEntityMultiBlock> test() {
        int length = 1;
        while (length < 65) {
            if (Reservoir.isRes(worldObj, pos.east(length), getType()))
                ++length;
            else
                break;
        }

        int width = 1;
        while (width < 65)
            if (Reservoir.isRes(worldObj, pos.south(width), getType()))
                ++width;
            else
                break;

        int height = 1;
        while (height < 65)
            if (Reservoir.isRes(worldObj, pos.up(height), getType()))
                ++height;
            else
                break;

        List<BlockPos> l1 = Reservoir.getNotHorizontalWall(worldObj, pos.getX(), pos.getY(), pos.getZ(), length, height, getType());
        if (l1.size() != 0) {
            size.valid = false;
            lastHeight = lastWidth = lastLength = -1;
            return null;
        }

        List<BlockPos> l2 = Reservoir.getNotHorizontalWall(worldObj, pos.getX(), pos.getY(), pos.getZ() + width - 1, length, height, getType());
        if (l2.size() != 0) {
            size.valid = false;
            lastHeight = lastWidth = lastLength = -1;
            return null;
        }

        List<BlockPos> l3 = Reservoir.getNotVerticalWall(worldObj, pos.getX(), pos.getY(), pos.getZ(), width, height, getType());
        if (l3.size() != 0) {
            size.valid = false;
            lastHeight = lastWidth = lastLength = -1;
            return null;
        }

        List<BlockPos> l4 = Reservoir.getNotVerticalWall(worldObj, pos.getX() + length - 1, pos.getY(), pos.getZ(), width, height, getType());
        if (l4.size() != 0) {
            size.valid = false;
            lastHeight = lastWidth = lastLength = -1;
            return null;
        }

        List<BlockPos> lfloor = Reservoir.getNotFloor(worldObj, pos.getX(), pos.getY(), pos.getZ(), length, width, getType());
        if (lfloor.size() != 0) {
            size.valid = false;
            lastHeight = lastWidth = lastLength = -1;
            return null;
        }

        lastHeight = height;
        lastWidth = width;
        lastLength = length;
        return clientSetMultiBlocks();
    }

    public ArrayList<TileEntityMultiBlock> clientSetMultiBlocks() {
        ArrayList<TileEntityMultiBlock> al = new ArrayList<TileEntityMultiBlock>();
        ArrayList<TileEntityMultiBlock> tmp;

        int length = lastLength, width = lastWidth, height = lastHeight;

        if (length == -1 || width == -1 || height == -1)
            return null;

        ocean.clear();
        tmp = Reservoir.getHorizontalWall(worldObj, pos.getX(), pos.getY(), pos.getZ(), length, height, getType());
        al.addAll(tmp);
        ocean.add(Reservoir.getHorizontalWallWater(worldObj, pos.getX(), pos.getY(), pos.getZ() - 1, length, height, getType()));
        tmp = Reservoir.getHorizontalWall(worldObj, pos.getX(), pos.getY(), pos.getZ() + width - 1, length, height, getType());
        al.addAll(tmp);
        ocean.add(Reservoir.getHorizontalWallWater(worldObj, pos.getX(), pos.getY(), pos.getZ() + width, length, height, getType()));
        tmp = Reservoir.getVerticalWall(worldObj, pos.getX(), pos.getY(), pos.getZ(), width, height, getType());
        al.addAll(tmp);
        ocean.add(Reservoir.getVerticalWallWater(worldObj, pos.getX() - 1, pos.getY(), pos.getZ(), width, height, getType()));
        tmp = Reservoir.getVerticalWall(worldObj, pos.getX() + length - 1, pos.getY(), pos.getZ(), width, height, getType());
        al.addAll(tmp);
        ocean.add(Reservoir.getVerticalWallWater(worldObj, pos.getX() + length, pos.getY(), pos.getZ(), width, height, getType()));
        tmp = Reservoir.getFloor(worldObj, pos.getX(), pos.getY(), pos.getZ(), length, width, getType());
        al.addAll(tmp);
        if (pos.getY() > 0)
            ocean.add(Reservoir.getFloorWater(worldObj, pos.getX(), pos.getY() - 1, pos.getZ(), length, width, getType()));

        size.valid = true;
        size.length = length;
        size.width = width;
        size.height = height;
        size.nonAirBlock = Reservoir.getNonAirBlock(worldObj, pos.getX(), pos.getY(), pos.getZ(), length, width, height);

        defaultStorage = size.getCapacity() * getType().capacity;
        setFluidTankCapacity(defaultStorage + extraStorage);

        return al;
    }

    @Override
    protected void onUpdate() {

        if (!isMaster() || WaterPower.isClientSide() || !size.valid)
            return;

        double[] ds = Utils.getBiomeRaining(worldObj, pos);
        int weather = (int) ds[0];
        double biomeGet = ds[1];
        double biomePut = ds[2];

        int length = size.getLength() - 2;
        int width = size.getWidth() - 2;
        int area = length * width;
        int cover = Reservoir.getCoverBlock(worldObj, pos.getX(), pos.getY() + size.getHeight(), pos.getZ(), size.getLength(), size.getWidth());

        double addWater = 0;
        double add = getType().capacity / 10000.0;
        // Rain Receiving
        addWater += rainLevel * weather * add * (area - cover) * biomeGet;
        // Tide Receiving
        addWater += Math.min(this.ocean.k, this.ocean.t) * getType().capacity / 100;
        // Underground water receiving
        addWater += underLevel * (1D - this.pos.getY() / 256) * add * (area - cover) * biomeGet;
        // Surface water receiving
        addWater += overLevel * (1D - Math.abs(64 - this.pos.getY()) / 64) * (area - cover) * biomeGet * add;

        String biomeID = worldObj.getBiomeGenForCoords(pos).getBiomeName();
        if (("ocean".equals(biomeID) || "river".equals(biomeID)) && pos.getY() < 64) {
            addWater += length * width * 0.5;
        }

        int delWater = (int) (weather == 0 ? area * 0.02 * biomePut : 0);

        preLastAddedWater = lastAddedWater;
        lastAddedWater = (int) ((addWater - delWater) * Reference.General.updateTick);

        if ((int) addWater * Reference.General.updateTick >= 1)
            fluidTank.fill(new FluidStack(FluidRegistry.WATER, (int) addWater * Reference.General.updateTick), true);
        fluidTank.drain(delWater * Reference.General.updateTick, true);
    }

    @Override
    public void update() {
        super.update();

        if (worldObj == null || worldObj.isRemote)
            return;

        boolean needsInvUpdate = false;

        if (needsFluid()) {
            MutableObject<ItemStack> output = new MutableObject<ItemStack>();

            if ((this.getFluidSlot().transferToTank(this.fluidTank, output, true))
                    && ((output.getValue() == null) || (this.outputSlot.canAdd(output.getValue())))) {
                needsInvUpdate = this.getFluidSlot().transferToTank(this.fluidTank, output, false);

                if (output.getValue() != null)
                    this.getOutputSlot().add(output.getValue());
            }

        }

        MutableObject<ItemStack> output = new MutableObject<ItemStack>();
        if ((this.getFluidSlot().transferFromTank(this.fluidTank, output, true))
                && ((output.getValue() == null) || (this.outputSlot.canAdd(output.getValue())))) {
            needsInvUpdate = this.getFluidSlot().transferFromTank(this.fluidTank, output, false);

            if (output.getValue() != null)
                this.getOutputSlot().add(output.getValue());
        }

        if (needsInvUpdate) {
            markDirty();
        }

        boolean flag = false;

        for (EnumFacing f : EnumFacing.values()) {
            TileEntity te = worldObj.getTileEntity(pos.offset(f));
            if (te instanceof IWaterReceiver) {
                IWaterReceiver te2 = (IWaterReceiver) te;
                int i = te2.canProvideWater(this.getFluidAmount(), f.getOpposite(), this);
                if (i > 0) {
                    te2.provideWater(i);
                    this.useLiquid(i);
                    flag = true;
                }
            }
        }

        if (flag)
            sendUpdateToClient();
    }

    @Override
    public int getGuiId() {
        return DefaultGuiIds.get("tileEntityReservoir");
    }

    @Override
    public boolean canFill(EnumFacing from, Fluid fluid) {
        // return
        // FluidRegistry.getFluidName(fluid.getID()).contentEquals("water");
        return true;
    }

    @Override
    public boolean canDrain(EnumFacing from, Fluid fluid) {
        // /return
        // FluidRegistry.getFluidName(fluid.getID()).contentEquals("water");
        return true;
    }

    @Override
    protected void onTestFailed() {
        setFluidTankCapacity(0);
    }

    public int getLastAddedWater() {
        if (isMaster())
            return lastAddedWater;
        else
            return ((TileEntityReservoir) masterBlock).getLastAddedWater();
    }

    @Override
    public void onLoaded() {
        if (isServerSide())
            refreshPlugins();

        super.onLoaded();
    }

    @Override
    public void markDirty() {
        super.markDirty();

        if (isServerSide())
            refreshPlugins();
    }

    public void refreshPlugins() {
        if (!isMaster())
            return;
        extraStorage = 0;
        rainLevel = overLevel = underLevel = 0;

        for (int i = 0; i < this.upgradeSlot.size(); i++) {
            ItemStack stack = this.upgradeSlot.get(i);

            if ((stack == null) || (!(stack.getItem() instanceof IUpgrade)))
                continue;
            IUpgrade upgrade = (IUpgrade) stack.getItem();
            rainLevel += upgrade.getRainAdditionalValue(stack);
            overLevel += upgrade.getOverworldAdditionalValue(stack);
            underLevel += upgrade.getUnderworldAdditionalValue(stack);
            extraStorage += upgrade.getStorageAdditionalValue(stack);
        }

        setFluidTankCapacity(defaultStorage + extraStorage);
    }

    @Override
    protected boolean allowedSendPacketTank() {
        return isMaster();
    }
}

File: src/main/kotlin/waterpower/common/block/reservoir/TileEntityReservoir.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.reservoir

import net.minecraft.util.EnumFacing
import net.minecraftforge.fluids.FluidRegistry
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.FluidTank
import net.minecraftforge.fluids.capability.IFluidHandler
import waterpower.Preference
import waterpower.annotations.HasGui
import waterpower.annotations.Register
import waterpower.annotations.SaveNBT
import waterpower.annotations.Sync
import waterpower.api.IUpgrade
import waterpower.api.IWaterReceiver
import waterpower.api.WaterAPI
import waterpower.common.block.attachment.AttachmentFluid
import waterpower.common.block.inventory.InventorySlotConsumableLiquid
import waterpower.common.block.inventory.InventorySlotOutput
import waterpower.common.block.inventory.InventorySlotUpgrade
import waterpower.common.block.tile.TileEntityMultiBlock
import waterpower.common.init.WPBlocks
import waterpower.util.Vec2i
import waterpower.util.getWaterIncomeAndExpenseByBiome

@HasGui(guiClass = GuiReservoir::class, containerClass = ContainerReservoir::class)
open class TileEntityReservoir(val type: Reservoirs) : TileEntityMultiBlock<TileEntityReservoir>(), IFluidHandler {

    protected val fluid = addAttachment(AttachmentFluid(this))
    @SaveNBT @Sync
    protected val fluidTank = fluid.addTank("fluid", 0)

    @Sync val reservoir = Reservoir()

    var defaultStorage = 0
    var extraStorage = 0

    var rainLevel = 0.0
    var underLevel = 0.0
    var overLevel = 0.0

    @Sync private var lastAddedWater = 0

    private val ocean = Vec2i()
    private val fluidSlot = InventorySlotConsumableLiquid(this, "fluidSlot", 1)
    private val outputSlot = InventorySlotOutput(this, "output", 1)
    private val upgradeSlot = InventorySlotUpgrade(this, "upgrade", 4)

    private lateinit var helper: MultiBlockHelper

    fun getFluidTank(): FluidTank {
        if (!isNormalBlock())
            return fluidTank
        else
            return getMasterBlock()!!.fluidTank
    }

    fun getFluidSlot(): InventorySlotConsumableLiquid {
        if (!isNormalBlock())
            return fluidSlot
        else
            return getMasterBlock()!!.fluidSlot
    }

    fun getOutputSlot(): InventorySlotOutput {
        if (!isNormalBlock())
            return outputSlot
        else
            return getMasterBlock()!!.outputSlot
    }

    fun getUpgradeSlot(): InventorySlotUpgrade {
        if (!isNormalBlock())
            return upgradeSlot
        else
            return getMasterBlock()!!.upgradeSlot
    }

    override fun getName() = type.getLocalizedName()

    override fun getBlockState() = super.getBlockState().withProperty(WPBlocks.reservoir.TYPES, type)!!

    fun useLiquid(use: Int) {
        if (getWorld().isRemote || isMaster())
            fluidTank.drain(use, true)
        else if (!isIncompleteStructure())
            (getMasterBlock() as TileEntityReservoir).fluidTank.drain(use, true)
    }

    override fun canBeMaster() = helper.canBeMaster()

    override fun onMasterChanged() {
        super.onMasterChanged()

        if (getMasterBlock() != this) {
            fluidTank.capacity = 0
            fluidTank.fluid = null
        }
    }

    override fun test(): List<TileEntityReservoir> {
        helper.reservoir(reservoir)

        if (reservoir.uuid != null) {
            defaultStorage = reservoir.capacity * type.capacity
            fluidTank.capacity = defaultStorage + extraStorage
        }

        return helper.list
    }

    override fun onUpdate() {
        super.onUpdate()

        if (isMaster() && !getWorld().isRemote && reservoir.uuid != null) {

            val (weather, biomeGet) = getWaterIncomeAndExpenseByBiome(getWorld(), pos)

            val length = reservoir.length - 2
            val width = reservoir.width - 2
            val area = length * width
            val cover = helper.coveredBlocks(reservoir)

            var addWater = 0.0
            val add = type.capacity / 10000.0
            // Rain Receiving
            addWater += rainLevel * weather * add * (area - cover) * biomeGet
            // Tide Receiving
            addWater += Math.min(this.ocean.x, this.ocean.y) * type.capacity / 100
            // Underground water receiving
            addWater += underLevel * (1.0 - this.pos.y / 256) * add * (area - cover) * biomeGet
            // Surface water receiving
            addWater += overLevel * (1.0 - Math.abs(64 - this.pos.y) / 64) * (area - cover) * biomeGet * add

            val biomeID = getWorld().getBiomeForCoordsBody(pos).biomeName
            if (("ocean" == biomeID || "river" == biomeID) && pos.y < 64) {
                addWater += length.toDouble() * width.toDouble() * 0.5
            }

            lastAddedWater = (addWater * Preference.General.updateTicks).toInt()

            if (addWater.toInt() * Preference.General.updateTicks >= 1)
                fluidTank.fill(FluidStack(FluidRegistry.WATER, addWater.toInt() * Preference.General.updateTicks), true)
        }
    }

    override fun update() {
        super.update()

        if (!getWorld().isRemote && !isIncompleteStructure()) {

            if (isMaster()) {
                if (fluidSlot.processIntoTank(fluidTank, outputSlot))
                    markDirty()
            }

            for (f in EnumFacing.values()) {
                val te = getWorld().getTileEntity(pos.offset(f))
                if (te != null && te.hasCapability(WaterAPI.CAPABILITY_WATER_RECEIVER!!, f.opposite)) {
                    val waterReceiver = te.getCapability(WaterAPI.CAPABILITY_WATER_RECEIVER!!, f.opposite) as IWaterReceiver
                    val delta = waterReceiver.receiveWater(getFluidTank().fluidAmount, f.opposite, false)
                    useLiquid(delta)
                }
            }
        }
    }

    override fun onTestFailed() {
        fluidTank.capacity = 0
    }

    fun getLastAddedWater(): Int {
        if (!isNormalBlock())
            return lastAddedWater
        else
            return (getMasterBlock() as TileEntityReservoir).getLastAddedWater()
    }

    override fun markDirty() {
        super.markDirty()

        if (isServerSide())
            refreshPlugins()
    }

    override fun onFirstTick() {
        super.onFirstTick()

        if (isServerSide()) {
            helper = MultiBlockHelper(this)
            refreshPlugins()
        }
    }

    fun refreshPlugins() {
        if (isMaster()) {
            extraStorage = 0
            underLevel = 0.0
            overLevel = 0.0
            rainLevel = 0.0

            for (i in 0..this.upgradeSlot.size() - 1) {
                val stack = this.upgradeSlot[i]

                if (stack.isEmpty || stack.item !is IUpgrade)
                    continue
                val upgrade = stack.item as IUpgrade
                rainLevel += upgrade.getRainAdditionalValue(stack)
                overLevel += upgrade.getOverworldAdditionalValue(stack)
                underLevel += upgrade.getUnderworldAdditionalValue(stack)
                extraStorage += upgrade.getStorageAdditionalValue(stack)
            }

            fluidTank.capacity = defaultStorage + extraStorage
        }
    }

    override fun drain(resource: FluidStack?, doDrain: Boolean) = getFluidTank().drain(resource, doDrain)

    override fun drain(maxDrain: Int, doDrain: Boolean) = getFluidTank().drain(maxDrain, doDrain)

    override fun fill(resource: FluidStack?, doFill: Boolean) = getFluidTank().fill(resource, doFill)

    override fun getTankProperties() = getFluidTank().tankProperties!!
}

@Register("waterpower.reservoir.wood")
class TileEntityReservoirWood : TileEntityReservoir(Reservoirs.wood)

@Register("waterpower.reservoir.stone")
class TileEntityReservoirStone : TileEntityReservoir(Reservoirs.stone)

@Register("waterpower.reservoir.lapis")
class TileEntityReservoirLapis : TileEntityReservoir(Reservoirs.lapis)

@Register("waterpower.reservoir.tin")
class TileEntityReservoirTin : TileEntityReservoir(Reservoirs.tin)

@Register("waterpower.reservoir.copper")
class TileEntityReservoirCopper : TileEntityReservoir(Reservoirs.copper)

@Register("waterpower.reservoir.quartz")
class TileEntityReservoirQuartz : TileEntityReservoir(Reservoirs.quartz)

@Register("waterpower.reservoir.zinc_alloy")
class TileEntityReservoirZincAlloy : TileEntityReservoir(Reservoirs.zinc_alloy)

@Register("waterpower.reservoir.bronze")
class TileEntityReservoirBronze : TileEntityReservoir(Reservoirs.bronze)

@Register("waterpower.reservoir.iron")
class TileEntityReservoirIron : TileEntityReservoir(Reservoirs.iron)

@Register("waterpower.reservoir.obsidian")
class TileEntityReservoirObsidian : TileEntityReservoir(Reservoirs.obsidian)

@Register("waterpower.reservoir.steel")
class TileEntityReservoirSteel : TileEntityReservoir(Reservoirs.steel)

@Register("waterpower.reservoir.gold")
class TileEntityReservoirGold : TileEntityReservoir(Reservoirs.gold)

@Register("waterpower.reservoir.manganese_steel")
class TileEntityReservoirManganeseSteel : TileEntityReservoir(Reservoirs.manganese_steel)

@Register("waterpower.reservoir.diamond")
class TileEntityReservoirDiamond : TileEntityReservoir(Reservoirs.diamond)

@Register("waterpower.reservoir.vanadium_steel")
class TileEntityReservoirVanadiumSteel : TileEntityReservoir(Reservoirs.vanadium_steel)
